Impossible to expect an allocator for getter

If I try to expect an allocator in a getter arguments, I have a compilation error:
```
Compile Log Output:
@as(*const [40:0]u8, "getter should have only 1 argument: self"), @as(type, fn(*dom.html_collection.HTMLCollection, mem.Allocator) u32)
src/main_wpt.zig:67:27: error: function called at runtime cannot return value at comptime
    const apis = comptime try jsruntime.compile(DOM.Interfaces);
```

I suppose the same issue happens w/ setter too.

--- a/src/engines/v8/generate.zig
+++ b/src/engines/v8/generate.zig
@@ -494,8 +494,9 @@ fn generateGetter(
 
             // TODO: check func params length
 
-            // retrieve the zig object
             var args: func.args_T = undefined;
+
+            // retrieve the zig object
 
             if (!comptime T_refl.isEmpty()) {
                 const obj_ptr = getNativeObject(T_refl, all_T, info.getThis()) catch unreachable;
@@ -505,6 +506,11 @@ fn generateGetter(
                 } else if (self_T == *T_refl.Self()) {
                     @field(args, "0") = obj_ptr;
                 }
+            }
+
+            // provide allocator if needed
+            if ((func.args.len) == 2) {
+                @field(args, "1") == utils.allocator;
             }
 
             // call the corresponding zig object method
@@ -546,9 +552,11 @@ fn generateSetter(
 
             // TODO: check func params length
 
+            var args: func.args_T = undefined;
+
             // get the value set in javascript
             const js_value = v8.Value{ .handle = raw_value.? };
-            const arg_T = func.args[0];
+            const arg_T = func.args[func.index_offset];
             var zig_value: arg_T.T = undefined;
 
             if (arg_T.isNative()) {
@@ -573,13 +581,25 @@ fn generateSetter(
                     isolate.getCurrentContext(),
                 ) catch unreachable; // TODO: throw js exception
             }
+            const js_value_name = func.args[func.index_offset].name.?;
+            @field(args, js_value_name) = zig_value;
 
             // retrieve the zig object
-            const obj_ptr = getNativeObject(T_refl, all_T, info.getThis()) catch unreachable;
+            const obj_ptr = getNativeObject(
+                T_refl,
+                all_T,
+                info.getThis(),
+            ) catch unreachable;
+            @field(args, "0") = obj_ptr;
+
+            // provide allocator if needed
+            if ((func.args.len) == 3) {
+                @field(args, "1") == utils.allocator;
+            }
 
             // call the corresponding zig object method
             const setter_func = @field(T_refl.T, func.name);
-            _ = @call(.auto, setter_func, .{ obj_ptr, zig_value }); // return should be void
+            _ = @call(.auto, setter_func, args); // return should be void
 
             // return to javascript the provided value
             info.getReturnValue().setValueHandle(raw_value.?);

--- a/src/reflect.zig
+++ b/src/reflect.zig
@@ -420,18 +420,13 @@ pub const Func = struct {
             @panic("func is not a function");
         }
 
-        // check args length
+        // check self is present
         var args = func.Fn.params;
         if (kind != .constructor and args.len == 0) {
             // TODO: handle "class methods"
             const msg = "getter/setter/methods should have at least 1 argument, self";
             fmtErr(msg.len, msg, T);
             return error.FuncNoSelf;
-        }
-        if (kind == .getter and args.len > 1) {
-            const msg = "getter should have only 1 argument: self";
-            fmtErr(msg.len, msg, T);
-            return error.FuncGetterMultiArg;
         }
 
         // self special case (only for methods)
@@ -515,6 +510,33 @@ pub const Func = struct {
             if (args_types[i].underErr() != null) {
                 return error.FuncErrorUnionArg;
             }
+        }
+
+        // check getter and setter internal argument is an allocator
+        if (kind == .getter or kind == .setter) {
+            if (index_offset == 1) {
+                if (args[0].type.? != std.mem.Allocator) {
+                    const msg = "getter/setter non-internal argument should be an allocator";
+                    fmtErr(msg.len, msg, T);
+                    return error.FuncGetterSetterNotAllocator;
+                }
+            }
+        }
+
+        const js_args_nb = args.len - index_offset;
+
+        // check getter has no js arg
+        if (kind == .getter and js_args_nb > 0) {
+            const msg = "getter should have only 1 JS argument: self";
+            fmtErr(msg.len, msg, T);
+            return error.FuncGetterMultiArg;
+        }
+
+        // check setter has at least one js arg
+        if (kind == .setter and js_args_nb == 0) {
+            const msg = "setter should have 1 JS argument";
+            fmtErr(msg.len, msg, T);
+            return error.FuncSetterNoArg;
         }
 
         // first optional arg
@@ -1284,8 +1306,10 @@ const Error = error{
 
     // func errors
     FuncNoSelf,
+    FuncGetterSetterNotAllocator,
     FuncGetterMultiArg,
     FuncSetterFirstArgNotSelfPtr,
+    FuncSetterNoArg,
     FuncGetterMethodFirstArgNotSelfOrSelfPtr,
     FuncVoidArg,
     FuncMultiCbk,
@@ -1380,10 +1404,18 @@ const TestFuncNoSelf = struct {
     pub fn _example() void {}
 };
 const TestFuncGetterMultiArg = struct {
-    pub fn get_example(_: TestFuncGetterMultiArg, _: anytype) void {}
+    pub fn get_example(_: TestFuncGetterMultiArg, _: bool) void {}
+};
+const TestFuncGetterSetterNotAllocator = struct {
+    pub fn get_example(_: TestFuncGetterSetterNotAllocator, _: *Loop) bool {
+        return true;
+    }
 };
 const TestFuncSetterFirstArgNotSelfPtr = struct {
     pub fn set_example(_: TestFuncSetterFirstArgNotSelfPtr) void {}
+};
+const TestFuncSetterNoArg = struct {
+    pub fn set_example(_: *TestFuncSetterNoArg) void {}
 };
 const TestFuncGetterMethodFirstArgNotSelfOrSelfPtr = struct {
     pub fn _example(_: bool) void {}
@@ -1491,8 +1523,16 @@ pub fn tests() !void {
         error.FuncGetterMultiArg,
     );
     try ensureErr(
+        .{TestFuncGetterSetterNotAllocator},
+        error.FuncGetterSetterNotAllocator,
+    );
+    try ensureErr(
         .{TestFuncSetterFirstArgNotSelfPtr},
         error.FuncSetterFirstArgNotSelfPtr,
+    );
+    try ensureErr(
+        .{TestFuncSetterNoArg},
+        error.FuncSetterNoArg,
     );
     try ensureErr(
         .{TestFuncGetterMethodFirstArgNotSelfOrSelfPtr},

--- a/src/tests/proto_test.zig
+++ b/src/tests/proto_test.zig
@@ -33,6 +33,14 @@ const Person = struct {
         return self.age;
     }
 
+    pub fn get_allocator(_: Person, _: std.mem.Allocator) bool {
+        return true;
+    }
+
+    pub fn set_allocator(_: *Person, _: std.mem.Allocator, _: bool) bool {
+        return true;
+    }
+
     pub fn set_age(self: *Person, age: u32) void {
         self.age = age;
     }
@@ -224,6 +232,7 @@ pub fn exec(
     // 2. getter
     var cases2 = [_]tests.Case{
         .{ .src = "p.age === 40", .ex = "true" },
+        .{ .src = "p.allocator", .ex = "true" },
     };
     try tests.checkCases(js_env, &cases2);
 
@@ -231,6 +240,7 @@ pub fn exec(
     var cases3 = [_]tests.Case{
         .{ .src = "p.age = 41;", .ex = "41" },
         .{ .src = "p.age", .ex = "41" },
+        .{ .src = "p.allocator = true", .ex = "true" },
     };
     try tests.checkCases(js_env, &cases3);
 

